src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -0,0 +1,261 @@
+                    if (connector is not null)
+                    {
+                        if (!validHostFunc(pool, preferedType))
+                        {
+                            conn.Connector = null;
+                            connector.Connection = null;
+                            pool.Return(connector);
+                            continue;
+                        }
+
+                        conn.Connector = connector;
+                        connector.Connection = conn;
+                        return connector;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    exceptions.Add(new NpgsqlException($"Unable to connect to {pool.Settings.Host}:{pool.Settings.Port}", ex));
+                }
+            }
+
+            return null;
+        }
+
+        async ValueTask<NpgsqlConnector?> TryRent(NpgsqlConnection conn, TimeSpan timeoutPerHost, bool async, TargetServerType preferedType,
+            Func<ConnectorPool, TargetServerType, bool> validHostFunc, IList<Exception> exceptions,
+            CancellationToken cancellationToken)
+        {
+            foreach (var pool in _pools)
+            {
+                if (!validHostFunc(pool, preferedType))
+                    continue;
+
+                NpgsqlConnector? connector = null;
+
+                try
+                {
+                    connector = await pool.Rent(conn, new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+                    if (GetHostState(pool) == HostState.Unknown)
+                        await connector.UpdateServerType(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+
+                    if (!validHostFunc(pool, preferedType))
+                    {
+                        conn.Connector = null;
+                        connector.Connection = null;
+                        pool.Return(connector);
+                        continue;
+                    }
+
+                    conn.Connector = connector;
+                    connector.Connection = conn;
+                    return connector;
+                }
+                catch (Exception ex)
+                {
+                    if (connector is not null)
+                    {
+                        conn.Connector = null;
+                        connector.Connection = null;
+                        pool.Return(connector);
+                    }
+
+                    exceptions.Add(new NpgsqlException($"Unable to connect to {pool.Settings.Host}:{pool.Settings.Port}", ex));
+                }
+            }
+
+            return null;
+        }
+
+        internal override async ValueTask<NpgsqlConnector> Rent(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken)
+        {
+            var exceptions = new List<Exception>();
+
+            var timeoutPerHost = timeout.IsSet ? timeout.CheckAndGetTimeLeft() : TimeSpan.Zero;
```

Comment:
This means a apply the timeout separately for each host. I'm not sure if that's a good idea - I think the expectation is for the timeout to be global for the connection attempt. The current multi-IP mechanism in NpgsqlConnector.Connect does this - the total timeout is divided by the number of IPs corresponding to the hostname.

But it's a complicated question, let's discuss.

Reply:
There is also a problem that we do not know how many hosts we will try. It might be every single one, or just one.

Reply:
Yeah. We can also split this out to a separate issue if you want.

Reply:
IMO that's actually fine as is, as with failover you actually care more about connecting to anything (that's why there is multiple hosts), rather than fail with a timeout.

Reply:
I'm not so sure.. I tend to think of the Timeout parameter as a contract for how long Open/OpenAsync will wait before throwing - regardless of what happens inside there. This kind of guarantee can be important in certain cases.

I don't have a clear opinion on this yet, but we can definitely defer the discussion.

Reply:
Hmm, for the use-cases I have we'd add `server1,server2` and it's _very likely_ that if `server2` is the master, then `server1` is actually down _and_ that we'd want that to behave "normally" ad-infinitum. As in the ideal world we wouldn't want to incur additional downtime to flip the master back over or have to go round and change all the connectionstrings etc.

Reply:
@mintsoft I'm not totally clear on what exactly you're saying with regards to the timeout being per-host or global for the entire open attempt...

To be clear, this interacts with the question of opening connections in parallel or in serial (#3610). If we open connections in serial, then is should the Timeout connection string parameter apply for each host separately, or to the entire attempt as a whole?

If we open connections in parallel, then it seems the Timeout question indeed no longer exists (assuming we open *all* connections in parallel).

Reply:
Sorry, I'll clarify. Personally I'd expect the timeout to be global for the entire attempt. I think the timeout should represent the maximum duration that the client application will wait for until either getting a successful connection or a `TimeoutException`